ietf/doc/forms.py

```diff
@@ -87,24 +87,39 @@ class NotifyForm(forms.Form):
     def clean_notify(self):
         # As long as the widget is a Textarea, users will separate addresses with newlines, whether that matches the instructions or not
         # We have been allowing nameaddrs for a long time (there are many Documents with namaddrs in their notify field)
-        nameaddrs = set([x.strip() for x in self.cleaned_data["notify"].replace('\n', ',').split(',')])
-        nameaddrs.discard('')
+        # python set doesn't preserve order, so in an attempt to mostly preserve the order of what was entered, we'll use
+        # a dict (whose keys are guaranteed to be ordered) to cull out duplicates
+
+        nameaddrs=dict()
+        duplicate_nameaddrs = set()
+        for nameaddr in self.cleaned_data["notify"].replace("\n", ",").split(","):
+            stripped = nameaddr.strip()
+            if stripped != "":
+                if stripped in nameaddrs:
+                    duplicate_nameaddrs.add(stripped)
+                else:
+                    nameaddrs[stripped] = None # only using this dict for its ordered keys
         bad_nameaddrs = []
-        for nameaddr in nameaddrs:
-            if '<' in nameaddr:
-                if nameaddr[-1] != '>':
+        for nameaddr in nameaddrs.keys():
+            if "<" in nameaddr:
+                if nameaddr[-1] != ">":
                     bad_nameaddrs.append(nameaddr)
                     continue
-                addrspec = nameaddr[nameaddr.find('<')+1:-1]
+                addrspec = nameaddr[nameaddr.find("<")+1:-1]
             else:
                 addrspec = nameaddr
             try:
                 validate_email(addrspec)
             except ValidationError:
                 bad_nameaddrs.append(nameaddr)
+        error_messages = []
+        if len(duplicate_nameaddrs) != 0:
+            error_messages.append(f'Duplicate addresses: {", ".join(duplicate_nameaddrs)}')
```

Comment:
Is there a need to make the user manually remove duplicates rather than simply dropping one? I guess maybe that isn't desirable if the order is important.

Maybe more of a feature request than a review comment, but I wonder if it'd be more helpful to filter for duplicates on `addrspec` rather than `nameaddr`. Having both `John Smith <js@example.com>` and `Smith, J <js@example.com>` is a pretty common dup error for mailing lists. If we're going to insist the user fix duplicates, would this be more helpful to report?

Reply:
I originally automatically just removed duplicates, but in manually testing it found that it was surprising to put a new address on the end (not seeing that it was already there) and having it not show up on the end after the form supposedly validated and saved. Since we're letting these strings get really long, calling out duplicates rather than quietly burying them seemed the right thing to do.

Reply:
You are right - the duplication concern is really at addrspec - will adjust.